src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -4445,7 +4453,26 @@ FindReferencedTableColumn(Expr *columnExpression, List *parentQueryList, Query *
 	 */
 	if (candidateColumn->varlevelsup > 0)
 	{
-		return;
+		int parentQueryIndex = list_length(parentQueryList) -
+							   candidateColumn->varlevelsup;
+		if (skipOuterVars || !(IsIndexInRange(parentQueryList, parentQueryIndex)))
+		{
+			return;
+		}
+
+		/*
+		 * Before we recurse into the query tree, we should update the candidateColumn and we use copy of it.
+		 * As we get the query from varlevelsup up, we reset the varlevelsup.
+		 */
+		candidateColumn = copyObject(candidateColumn);
+		candidateColumn->varlevelsup = 0;
+
+		/*
+		 * We should be careful about these fields because they need to
+		 * be updated correctly based on ctelevelsup and varlevelsup.
+		 */
+		query = list_nth(parentQueryList, parentQueryIndex);
+		parentQueryList = list_truncate(parentQueryList, parentQueryIndex);
```

Comment:
Lets add the following assert/comment (or something similar) to CTE section, because I really doubt this would work there. I think without supporting it, there is no code-base to test, but we should be careful when we enable in the future.

```C
	else if (rangeTableEntry->rtekind == RTE_CTE)
	{
		/*
		 * When outerVars are considered, we modify parentQueryList, so this
		 * logic might need to change when we support outervars in CTEs.
		 */
		Assert (!skipOuterVars);
```